src/command/handlers/http/undici.ts

```diff
@@ -0,0 +1,469 @@
+					},
+					keyType: cert.asn1Curve || cert.nistCurve ? 'EC' : cert.modulus || cert.exponent ? 'RSA' : null,
+					keyBits: cert.bits || null,
+					serialNumber: cert.serialNumber?.match(/.{2}/g)?.join(':') ?? null,
+					fingerprint256: cert.fingerprint256,
+					publicKey: cert.pubkey?.toString('hex').toUpperCase().match(/.{2}/g)?.join(':') ?? null,
+				};
+
+				callback(null, tlsSocket);
+			});
+		});
+
+		return tcpSocket;
+	};
+}
+
+export class Test {
+	private readonly REQUEST_TIMEOUT: number = 10_000;
+	private readonly DOWNLOAD_LIMIT: number = 10_000;
+	private readonly url: URL;
+	private readonly port: number;
+	private readonly isHttps: boolean;
+	private undiciClient!: Client;
+	private readonly result: Omit<Result, 'timings'>;
+	private readonly timings: Timings = {
+		start: null,
+		total: null,
+		dns: null,
+		tcp: null,
+		tls: null,
+		firstByte: null,
+		download: null,
+	};
+
+	private resolve!: (value: unknown) => void;
+	private timeoutTimer: NodeJS.Timeout | null = null;
+	private decompressor: Decompressor | null = null;
+	private decompressorHasData = false;
+	private done = false;
+
+	constructor (
+		private readonly options: HttpOptions,
+		private readonly buffer: ProgressBuffer,
+	) {
+		this.result = this.getInitialResult();
+		this.port = options.port ? options.port : (options.protocol === 'HTTP' ? 80 : 443);
+		this.isHttps = options.protocol !== 'HTTP';
+		this.url = new URL(this.urlBuilder());
+	}
+
+	public async run () {
+		const promise = new Promise((resolve) => { this.resolve = resolve; });
+		const dnsResolver = callbackify(dnsLookup(this.options.resolver), true);
+		const allowH2 = this.options.protocol === 'HTTP2';
+		const connector = getConnector(this.options, this.port, this.isHttps, dnsResolver, this.result, this.timings);
+		this.undiciClient = new Client(this.url.origin, { connect: connector, allowH2 });
+		this.timeoutTimer = setTimeout(() => this.handleError('Request timeout.'), this.REQUEST_TIMEOUT);
+
+		this.undiciClient.dispatch({
+			path: this.url.pathname + this.url.search,
+			method: this.options.request.method as Dispatcher.HttpMethod,
+			headers: lowerCaseKeys({
+				'Accept-Encoding': `gzip, deflate, br${createZstdDecompress ? ', zstd' : ''}`,
+				...this.options.request.headers,
+				'User-Agent': 'globalping probe (https://github.com/jsdelivr/globalping)',
+				'Host': this.options.request.host ?? this.options.target,
+				'Connection': 'close',
+			}),
+		}, {
+			onConnect: () => {},
+			onError: (err: Error) => this.handleError(err.message),
+			onHeaders: (statusCode, headers, _resume, statusText) => {
+				this.timings.firstByte = Date.now() - this.timings.start! - this.timings.dns! - this.timings.tcp! - this.timings.tls!;
+				this.result.statusCode = statusCode;
+				this.result.statusCodeName = statusText;
+
+				const rawHeaderPairs = [];
+
+				if (headers) {
+					const entries: [string, string][] = Array.isArray(headers)
```

Comment:
It seems we might be able to use `Headers` object (from undici) to handle most of this?

Reply:
Headers object from undici has lowecased keys, so there is no way to get rawHeaders from it.